app/support/idempotency.rb

```diff
@@ -0,0 +1,98 @@
+# frozen_string_literal: true
+
+# Usage:
+#
+#   # Execute an operation idempotently
+#   Idempotency.transaction do
+#     Idempotency.unique!(invoice, invoice.date, invoice.customer_id)
+
+#     # Perform your business logic here
+#     result = perform_operation
+#   end
+class Idempotency
+  # Thread-local storage for the current transaction
+  thread_mattr_accessor :current_transaction
+
+  class IdempotencyError < StandardError; end
+
+  # Represents a transaction context for an idempotent operation
+  class Transaction
+    attr_accessor :idempotent_resources
+
+    def initialize
+      @idempotent_resources = Hash.new { |k, v| k[v] = [] }
```

Comment:
TIL! I can't believe I didn't know about this way of initializing Hash!

Reply:
ah yes, common pitfall is to do `Hash.new([])` which means the same array will be re-used for _all_ keys.